packer/linux/conf/bin/bk-configure-cloudwatch-agent.sh

```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# shellcheck disable=SC2094
+set -euo pipefail
+
+if [[ "${CLOUDWATCH_ENABLE_METRICS:-false}" == "true" ]]; then
+    cw_config="/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json"
+    cat <<<"$(jq \
+        --arg queue "$BUILDKITE_QUEUE" \
+        '. + {
+            metrics: {
+                metrics_collected: {
+                    mem: {measurement: ["mem_used_percent"], append_dimensions: {BuildkiteQueue: $queue}},
+                    disk: {measurement: ["used_percent"], resources: ["*"], append_dimensions: {BuildkiteQueue: $queue}}
+                },
+                append_dimensions: {
+                    AutoScalingGroupName: "${aws:AutoScalingGroupName}"
+                }
+            }
+        }' $cw_config)" >$cw_config
```

Comment:
so looks like this will append to https://github.com/buildkite/elastic-ci-stack-for-aws/blob/509802db39718c8e0623535804237fed35f6f516/packer/linux/conf/cloudwatch-agent/config.json so that the existing file doesn't get clobbered?

Reply:
Yes, it actually rewrites it. It's the same pattern used to [configure docker](https://github.com/buildkite/elastic-ci-stack-for-aws/blob/19278b86f3e08cce38f144811f9c9c0c253589fd/packer/linux/conf/bin/bk-configure-docker.sh#L18), albeit more complicated.

* The `jq` command will read the `$cw_config` file, append the metrics configuration block and output it (`jq [options] <jq filter> [file...]`). Where the jq filter is:
  - `.` identity filter
  - `{}`: object construction
  - `+`: addition operator between two objects. So takes the original configuration and merge the new object created:
     > Objects are added by merging, that is, inserting all the key-value pairs from both objects into a single combined object.
* This output is passed to the standard input cat with a `here-string` (`<<<`) using command substitution (`$(..)`). So the jq command runs in a subshell.
* `cat` is then overwriting the original config file

It's equivalent to:

```bash
new_config=$(jq <filter> $cw_config)
echo $new_config | cat > $cw_config
```
 
From what I understand, the combination of here-string and command substitution will use a temp file or memory to avoid any race condition while modifying the file in place.